site/src/hooks/usePaginatedQuery.ts

```diff
@@ -0,0 +1,443 @@
+  const hasPreviousPage =
+    totalRecords !== undefined &&
+    currentPage > 1 &&
+    currentPageOffset - limit < totalRecords;
+
+  const queryClient = useQueryClient();
+  const prefetchPage = useEffectEvent((newPage: number) => {
+    const options = getQueryOptionsFromPage(newPage);
+    return queryClient.prefetchQuery(options);
+  });
+
+  // Have to split hairs and sync on both the current page and the hasXPage
+  // variables, because the page can change immediately client-side, but the
+  // hasXPage values are derived from the server and won't always be immediately
+  // ready on the initial render
+  useEffect(() => {
+    if (hasNextPage) {
+      void prefetchPage(currentPage + 1);
+    }
+  }, [prefetchPage, currentPage, hasNextPage]);
+
+  useEffect(() => {
+    if (hasPreviousPage) {
+      void prefetchPage(currentPage - 1);
+    }
+  }, [prefetchPage, currentPage, hasPreviousPage]);
+
+  // Mainly here to catch user if they navigate to a page directly via URL;
+  // totalPages parameterized to insulate function from fetch status changes
+  const updatePageIfInvalid = useEffectEvent(async (totalPages: number) => {
+    // If totalPages is 0, that's a sign that the currentPage overshot, and the
+    // API returned a count of 0 because it didn't know how to process the query
+    let fixedTotalPages: number;
+    if (totalPages !== 0) {
+      fixedTotalPages = totalPages;
+    } else {
+      const firstPageOptions = getQueryOptionsFromPage(1);
+      try {
+        const firstPageResult = await queryClient.fetchQuery(firstPageOptions);
+        fixedTotalPages = Math.ceil(firstPageResult?.count ?? 0 / limit) || 1;
+      } catch (err) {
+        fixedTotalPages = 1;
+      }
+    }
+
+    const clamped = clamp(currentPage, 1, fixedTotalPages);
+    if (currentPage === clamped) {
+      return;
+    }
+
+    const withoutPage = getParamsWithoutPage(searchParams);
+    if (onInvalidPageChange === undefined) {
+      withoutPage.set(PAGE_NUMBER_PARAMS_KEY, String(clamped));
+      setSearchParams(withoutPage);
+    } else {
+      const params: InvalidPageParams = {
+        limit,
+        setSearchParams,
+        offset: currentPageOffset,
+        searchParams: withoutPage,
+        totalPages: fixedTotalPages,
+        pageNumber: currentPage,
+      };
+
+      onInvalidPageChange(params);
+    }
+  });
+
+  useEffect(() => {
+    if (!query.isFetching && totalPages !== undefined) {
+      void updatePageIfInvalid(totalPages);
+    }
+  }, [updatePageIfInvalid, query.isFetching, totalPages]);
+
+  const onPageChange = (newPage: number) => {
+    // Page 1 is the only page that can be safely navigated to without knowing
+    // totalPages; no reliance on server data for math calculations
+    if (totalPages === undefined && newPage !== 1) {
+      return;
+    }
+
+    const cleanedInput = clamp(Math.trunc(newPage), 1, totalPages ?? 1);
+    if (!Number.isInteger(cleanedInput) || cleanedInput <= 0) {
```

Comment:
what makes these checks necessary? the `trunc` and `clamp` should already guarantee these already?

Reply:
There's still one check that needs to be done for NaN values – they won't throw errors for any of the math calculations

But you're right – one of the checks doesn't apply anymore, and the other can just turn into `Number.isNaN`